### PR TITLE
Support new Perpdex interface

### PR DESCRIPTION
**Pull request by takami-ml**
Changes
- event indexing
- liquidation logic

Todo
- testnet scenario tests (only for local test. skip in CI)
  - step1: alice deposit -> trade -> withdraw
  - step2: bob trade big amount
  - step3: liquidate

### CODE REVIEW

File: src/liquidator.py [hunk 1]
```diff
@@ -23,45 +24,108 @@ def __init__(self, logger=None) -> None:
...
+            caller=self._user_account.address,
+            is_base_to_quote=base_share > 0,
+            is_exact_input=True,
+        )).call()
+
+        # try liquidate
+        reduction_rate = 0.8
+        amount = max(
+            base_share,
+            max_trade * reduction_rate,
```
Comment (richmanbtc) [inline src/liquidator.py:98]: - min?
- abs(base_share)?

File: src/liquidator.py [hunk 2]
```diff
@@ -23,45 +24,108 @@ def __init__(self, logger=None) -> None:
...
+                return False
+
+    def _liquidate_taker_position(self, trader, market):
+        base_share = self._perpdex_exchange.functions.getOpenPositionShare(trader, market).call()
+        max_trade = self._perpdex_exchange.functions.maxTrade(dict(
+            trader=trader,
+            market=market,
+            caller=self._user_account.address,
+            is_base_to_quote=base_share > 0,
+            is_exact_input=True,
```
Comment (richmanbtc) [inline src/liquidator.py:91]: It seems that is_exact_input must be equal to is_base_to_quote

File: src/liquidator.py [hunk 3]
```diff
@@ -23,45 +24,108 @@ def __init__(self, logger=None) -> None:
...
+        reduction_rate = 0.8
+        amount = max(
+            base_share,
+            max_trade * reduction_rate,
+        )
+        func = self._perpdex_exchange.functions.trade(dict(
+            trader=trader,
+            market=market,
+            is_base_to_quote=base_share > 0,
+            is_exact_input=True,
```
Comment (richmanbtc) [inline src/liquidator.py:104]: It seems that is_exact_input must be equal to is_base_to_quote

**Comment by takami-ml** [comment]
> It is also possible to launch a node with perpdex-contracts, deploy a contract, and test against that contract, as we do in the following repositories. I don't know if this method is good though
https://github.com/alphasea-dapp/alphasea-agent/blob/master/.github/workflows/python-app.yml#L24

Thank you. I will do this in another PR.

File: src/liquidator.py (hunk 1 above)
Comment (takami-ml) [reply src/liquidator.py]: fixed.

File: src/liquidator.py (hunk 2 above)
Comment (takami-ml) [reply src/liquidator.py]: fixed.

File: src/liquidator.py (hunk 3 above)
Comment (takami-ml) [reply src/liquidator.py]: fixed.

**Pull request merged**